Fix Infinite Loops in ExceptionsHelper#unwrap

original-brownbear opened the pull request:
* Keep track of all seen exceptions and break out on loops
* Closes #42340

original-brownbear committed:
Fix Infinite Loops in ExceptionsHelper#unwrap

* Keep track of all seen exceptions and break out on loops
* Closes #42340

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch reviewed (changes requested):
This implements similar but slightly different logic than `maybeError`.
Can we do it one way and even reuse code?

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 42340

original-brownbear committed:
CR: reuse existing approach

original-brownbear commented:
@ywelsch sure done in 832db83 :) Reused the existing approach, I think it's good enough to go 1024 (right?) steps and it's safer without the recursion.

jpountz reviewed (changes requested):
Maybe I'm being over cautious, but I'd like to collect exceptions that have been seen so far better than using a counter. The reason is that with the counter approach, given that we look at suppressed exceptions before causes, if there is a cycle in suppressed exceptions then we will never look at the causes.

Maybe we also need to improve naming in this class, we have one `unwrap` method that only looks at causes and another one that looks at both causes and suppressed exceptions, this got me a bit confused while reviewing.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 42340

original-brownbear committed:
CR: Correctly handle a mix of circular and non-circular exceptions

original-brownbear committed:
CR: don't use a limit on unwrapping exceptions

original-brownbear commented:
@jpountz thanks for spotting that. I added a test for the case you mentioned, removed the step limit, put the set of all visited Exceptions back into action and removed the recursion -> let me know what you think :)
Didn't do any renaming though, not sure we want to mix renaming (that also affect other code) with the changes here?

original-brownbear commented:
Jenkins run elasticsearch-ci/1

jpountz commented on server/src/main/java/org/elasticsearch/ExceptionsHelper.java:
```diff
@@ -246,33 +244,22 @@ public static boolean reThrowIfNotNull(@Nullable Throwable e) {
-        if (cause instanceof Error) {
-            return Optional.of((Error) cause);
+    @SuppressWarnings("unchecked")
+    private static <T extends Throwable> Optional<T> unwrap(Throwable cause, Predicate<Throwable> predicate) {
```
when I mentioned naming, I was thinking that this one could e.g. be called something like `unwrapCausesAndSuppressed` (ugly, maybe you can think of something better :) ), which shouldn't be an issue since it's both new and private?

original-brownbear replied:
Ah that makes sense :) I couldn't think of anything better though 850d507

jpountz reviewed (approved):
LGTM

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 42340

original-brownbear committed:
CR: renamed method

original-brownbear commented:
@jpountz thanks!
@ywelsch you're good with this one as well?

ywelsch reviewed (approved):
LGTM
